app/ui/client/lib/UserAction.js

```diff
@@ -0,0 +1,177 @@
+
+function handleStreamAction(activity, performingUsers, rid, username, extras) {
+	const id = extras?.tmid ? extras.tmid : rid;
+	const users = performingUsers.get(id) || {};
+
+	if (activity === true) {
+		clearTimeout(users[username]);
+		users[username] = setTimeout(function() {
+			const u = performingUsers.get(id);
+			delete u[username];
+			performingUsers.set(id, u);
+		}, timeout);
+	} else {
+		clearTimeout(users[username]);
+		delete users[username];
+	}
+
+	performingUsers.set(id, users);
+}
+
+export const UserAction = new class {
+	constructor() {
+		Tracker.autorun(() => Session.get('openedRoom') && this.addStream(Session.get('openedRoom')));
+	}
+
+	get selfActivity() { return selfActivity; }
+
+	addStream(rid) {
+		if (rooms[rid]) {
+			return;
+		}
+		rooms[rid] = function(username, activity, activityType, extras) {
+			const user = Meteor.users.findOne(Meteor.userId(), { fields: { name: 1, username: 1 } });
+			if (username === shownName(user)) {
+				return;
+			}
+			if (activityType === USER_RECORDING) {
+				handleStreamAction(activity, recordingUsers, rid, username, extras);
+			} else if (activityType === USER_UPLOADING) {
+				handleStreamAction(activity, uploadingUsers, rid, username, extras);
+			} else if (activityType === USER_TYPING) {
+				handleStreamAction(activity, typingUsers, rid, username, extras);
+			}
+		};
+		return Notifications.onRoom(rid, USER_ACTIVITY, rooms[rid]);
+	}
+
+	startPerformingAction(rid, activityType, timeouts, renews, extras) {
+		if (renews[rid]) {
+			return;
+		}
+		selfActivity[activityType] = true;
+
+		renews[rid] = setTimeout(() => {
+			clearTimeout(renews[rid]);
+			delete renews[rid];
+		}, renew);
+
+		startActivity(rid, activityType, extras);
+		if (timeouts[rid]) {
+			clearTimeout(timeouts[rid]);
+			delete timeouts[rid];
+		}
+
+		timeouts[rid] = setTimeout(() => this.stop(rid, activityType), timeout);
+		return timeouts[rid];
+	}
+
+	start(rid, activityType, extras = {}) {
+		if (activityType === USER_RECORDING) {
+			this.startPerformingAction(rid, USER_RECORDING, recordingTimeouts, recordingRenews, extras);
+		} else if (activityType === USER_TYPING) {
+			this.startPerformingAction(rid, USER_TYPING, typingTimeouts, typingRenews, extras);
+		} else if (activityType === USER_UPLOADING) {
+			this.startPerformingAction(rid, USER_UPLOADING, uploadingTimeouts, uploadingRenews, extras);
+		} else {
+			// console.log('activity type is invalid');
+		}
+	}
+
+	stopPerformingAction(rid, activityType, timeouts, renews, extras) {
+		selfActivity[activityType] = false;
+		if (timeouts[rid]) {
+			clearTimeout(timeouts[rid]);
+			delete timeouts[rid];
+		}
+		if (renews[rid]) {
+			clearTimeout(renews[rid]);
+			delete renews[rid];
+		}
+		return stopActivity(rid, activityType, extras);
+	}
+
+	stop(rid, activityType, extras) {
+		if (activityType === USER_RECORDING) {
```

Comment:
If you use a sinlge renew and timeout object you can eliminate these ifs and pass the activity type directly